### PR TITLE
Check run name as well as workflow name in Cylc install

### LINKED ISSUE #6246: validate full workflow ID
The workflow "name" is validated, the "run name" is not.

```console
$ cylc install generic -n f? -r r?
WorkflowFilesError: invalid workflow name 'f?' - can only contain: alphanumeric, `/`, `_`, `+`, `-`, `.`, `@`
$ cylc install generic -n f -r r?
INSTALLED f/r? from /net/home/h06/osanders/cylc-src/generic
```

### PR DESCRIPTION
**Pull request by wxtim**
Closes #6246 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: changes.d/6264.fix.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix bug where  failed to prevent invalid run names.
```
Comment (MetRonnie) [inline changes.d/6264.fix.md:1]:
```suggestion
Fix bug where `cylc install` failed to prevent invalid run names.
```

File: cylc/flow/install.py [hunk 2]
```diff
@@ -285,13 +285,18 @@ def install_workflow(
     source = Path(expand_path(source)).resolve()
     if not workflow_name:
         workflow_name = get_source_workflow_name(source)
-    validate_workflow_name(workflow_name, check_reserved_names=True)
     if run_name is not None:
+        validate_workflow_name(
+            workflow_name + '/' + run_name,
```
Comment (MetRonnie) [inline cylc/flow/install.py:290]:
```suggestion
            os.path.join(workflow_name, run_name),
```

File: cylc/flow/workflow_files.py [hunk 3]
```diff
@@ -831,31 +832,36 @@ def check_deprecation(path, warn=True, force_compat_mode=False):
 
 
 def validate_workflow_name(
-    name: str, check_reserved_names: bool = False
+    name: str,
+    check_reserved_names: bool = False,
+    being_validated: Literal['workflow', 'workflow and run'] = 'workflow'
```
Comment (MetRonnie) [inline cylc/flow/workflow_files.py:837]: I don't see the need for this argument; I would just keep it simple and drop any changes to this function

**Review by MetRonnie — COMMENTED** [review]
Lint failing btw

File: cylc/flow/workflow_files.py (hunk 3 above)
Comment (wxtim) [reply cylc/flow/workflow_files.py:837]: I get your point, but if a user does

```
> cylc install --run-name  foo?
```

Then "invalid workflow name" seems unhelpful as a message. 

I'd be happy to modify the error messages without changing the functioning though.

Comment (wxtim) [reply cylc/flow/workflow_files.py:837]: Nah, I agree, it's fine

File: cylc/flow/install.py [hunk 4]
```diff
@@ -285,13 +285,15 @@ def install_workflow(
...
     if not workflow_name:
         workflow_name = get_source_workflow_name(source)
-    validate_workflow_name(workflow_name, check_reserved_names=True)
     if run_name is not None:
+        validate_workflow_name(os.path.join(workflow_name, run_name))
         if len(Path(run_name).parts) != 1:
             raise WorkflowFilesError(
                 f'Run name cannot be a path. (You used {run_name})'
             )
         check_reserved_dir_names(run_name)
```
Comment (MetRonnie) [inline cylc/flow/install.py:294]: Still need to check reserved dir names on the whole path
```suggestion
        validate_workflow_name(
            os.path.join(workflow_name, run_name),
            check_reserved_names=True
        )
        if len(Path(run_name).parts) != 1:
            raise WorkflowFilesError(
                f'Run name cannot be a path. (You used {run_name})'
            )
```

**Review by MetRonnie — CHANGES_REQUESTED** [review]

File: cylc/flow/install.py (hunk 4 above)
Comment (MetRonnie) [reply cylc/flow/install.py]: @wxtim I think there ought to be a test that covers this point

Comment (MetRonnie) [reply cylc/flow/install.py]: I think there ought to be a test that covers this point

**Review by MetRonnie — CHANGES_REQUESTED** [review]
See comment above; requested a test

File: tests/unit/test_install.py [hunk 5]
```diff
@@ -542,3 +542,47 @@ def test_validate_source_dir(tmp_run_dir: Callable, tmp_src_dir: Callable):
...
+        # And that check fails first.
+        # '.service'
+    }
+    for name in reserved_names:
+        with pytest.raises(WorkflowFilesError, match=is_reserved):
+            install_workflow(src_dir, workflow_name='ok', run_name=name)
+        with pytest.raises(WorkflowFilesError, match=is_reserved):
+            install_workflow(src_dir, workflow_name=name)
+        with pytest.raises(WorkflowFilesError, match=is_reserved):
+            install_workflow(src_dir, workflow_name=name, run_name=name)
```
Comment (MetRonnie) [inline tests/unit/test_install.py:588]: This doesn't capture the point I raised. Needs:
```suggestion
            install_workflow(src_dir, workflow_name=name, run_name='ok')
```

**Review by MetRonnie — CHANGES_REQUESTED** [review]

**Review by MetRonnie — DISMISSED** [review]

File: cylc/flow/install.py [hunk 6]
```diff
@@ -285,13 +285,18 @@ def install_workflow(
...
     if run_name is not None:
         if len(Path(run_name).parts) != 1:
             raise WorkflowFilesError(
                 f'Run name cannot be a path. (You used {run_name})'
             )
+        validate_workflow_name(
+            os.path.join(workflow_name, run_name),
+            check_reserved_names=True
+        )
         check_reserved_dir_names(run_name)
```
Comment (oliver-sanders) [inline cylc/flow/install.py:297]: Isn't the subsequent `check_reserved_dir_names(run_name)` superfluous with `check_reserved_names=True`?

File: tests/unit/test_install.py [hunk 7]
```diff
@@ -542,3 +542,47 @@ def test_validate_source_dir(tmp_run_dir: Callable, tmp_src_dir: Callable):
...
+    src_dir: Path = tmp_src_dir('ludlow')
+    # It only has a workflow name:
+    with pytest.raises(WorkflowFilesError, match='can only contain'):
+        install_workflow(src_dir, workflow_name='foo?')
+    # It only has a run name:
+    with pytest.raises(WorkflowFilesError, match='can only contain'):
+        install_workflow(src_dir, run_name='foo?')
+    # It has a legal workflow name, but an invalid run name:
+    with pytest.raises(WorkflowFilesError, match='can only contain'):
+        install_workflow(src_dir, workflow_name='foo', run_name='bar?')
```
Comment (oliver-sanders) [inline tests/unit/test_install.py:560]: With these sorts of tests where we are testing for failure, it is often a good idea to test that the thing you are testing can actually succeed when the conditions are right.

This avoids the situation where the test setup is broken causing the action to fail with the expected error, but for the wrong reasons (though the `match` kwarg provides some protection against this).

File: cylc/flow/install.py (hunk 6 above)
Comment (MetRonnie) [reply cylc/flow/install.py:297]: Ah yes, good spot (my suggestion https://github.com/cylc/cylc-flow/pull/6264#discussion_r1692729604 removed this but was not applied directly)

**Review by MetRonnie — APPROVED** [review]

**Review by MetRonnie — CHANGES_REQUESTED** [review]
Ah, flake8 failing...

**Comment by MetRonnie** [comment]
Lint workflow still failing...

**Review by oliver-sanders — APPROVED** [review]

**Pull request merged**